llvm/lib/SYCLLowerIR/PropagateAspectsUsage.cpp

```diff
@@ -0,0 +1,344 @@
+    const AspectsSetTy &Aspects =
+        getAspectsFromType(OperandIt->getType(), Types);
+    Result.insert(Aspects.begin(), Aspects.end());
+  }
+
+  return Result;
+}
+
+using FunctionToAspectsMapTy = DenseMap<Function *, AspectsSetTy>;
+using CallGraphTy = DenseMap<Function *, SmallPtrSet<Function *, 8>>;
+
+void createUsedAspectsMetadataForFunctions(FunctionToAspectsMapTy &Map) {
+  for (auto &It : Map) {
+    AspectsSetTy &Aspects = It.second;
+    if (Aspects.empty())
+      continue;
+
+    Function *F = It.first;
+    LLVMContext &C = F->getContext();
+
+    SmallVector<Metadata *, 16> AspectsMetadata;
+    for (const auto &A : Aspects)
+      AspectsMetadata.push_back(ConstantAsMetadata::get(
+          ConstantInt::getSigned(Type::getInt32Ty(C), A)));
+
+    MDNode *MDN = MDNode::get(C, AspectsMetadata);
+    F->setMetadata("intel_used_aspects", MDN);
+  }
+}
+
+/// Propagates aspects from leaves up to the top of call graph.
+/// NB! Call graph corresponds to call graph of SYCL code which
+/// can't contain recursive calls. So there can't be loops in
+/// a call graph. But there can be path's intersections.
+void propagateAspectsThroughCG(Function *F, CallGraphTy &CG,
+                               FunctionToAspectsMapTy &AspectsMap,
+                               SmallPtrSet<const Function *, 16> &Visited) {
+  const auto It = CG.find(F);
+  if (It == CG.end())
+    return;
+
+  AspectsSetTy LocalAspects;
+  for (Function *Callee : It->second) {
+    if (Visited.insert(Callee).second)
+      propagateAspectsThroughCG(Callee, CG, AspectsMap, Visited);
+
+    const auto &CalleeAspects = AspectsMap[Callee];
+    LocalAspects.insert(CalleeAspects.begin(), CalleeAspects.end());
+  }
+
+  AspectsMap[F].insert(LocalAspects.begin(), LocalAspects.end());
+}
+
+/// Processes a function:
+///  - checks if return and argument types are using any aspects
+///  - checks if instructions are using any aspects
+///  - updates call graph information
+///
+void processFunction(Function &F, FunctionToAspectsMapTy &FunctionToAspects,
+                     TypeToAspectsMapTy &TypesWithAspects, CallGraphTy &CG) {
+  const AspectsSetTy RetTyAspects =
+      getAspectsFromType(F.getReturnType(), TypesWithAspects);
+  FunctionToAspects[&F].insert(RetTyAspects.begin(), RetTyAspects.end());
+  for (Argument &Arg : F.args()) {
+    const AspectsSetTy ArgAspects =
+        getAspectsFromType(Arg.getType(), TypesWithAspects);
+    FunctionToAspects[&F].insert(ArgAspects.begin(), ArgAspects.end());
+  }
+
+  for (Instruction &I : instructions(F)) {
+    const AspectsSetTy Aspects =
+        getAspectsUsedByInstruction(I, TypesWithAspects);
+    FunctionToAspects[&F].insert(Aspects.begin(), Aspects.end());
+
+    if (const auto *CI = dyn_cast<CallInst>(&I)) {
+      if (!CI->isIndirectCall() && CI->getCalledFunction())
+        CG[&F].insert(CI->getCalledFunction());
+    }
+  }
+}
+
+// Return true if the function is a SPIRV or SYCL builtin, e.g.
```

Comment:
May be a comment here saying this (and the next function) code is shared with ModuleSplitter.cpp and mention that the two need to be in sync?

Reply:
Added such comment in 96139d671baf3ea32672932b73d2dc5c0a56663b